File: packages/swt3-ai/swt3_ai/deployment.py
"""SWT3 AI Witness SDK -- Deployment Context Detection.

Auto-detects cloud provider, region, runtime environment, and
accelerator hardware from environment variables and metadata.
Results are embedded in witness payloads as deployment_context.

Detection is best-effort and never blocks. Unknown fields remain
as "unknown". All raw identifiers are hashing-safe (no secrets).

Security: Container IDs and hostnames are SHA-256 hashed before
inclusion in any witness payload.
"""

from __future__ import annotations

import os
import time
from dataclasses import dataclass, asdict
from typing import Any, Dict, Optional

from .fingerprint import sha256_truncated


@dataclass
class DeploymentContext:
    cloud_provider: str = "unknown"
    region: str = "unknown"
    availability_zone: str = "unknown"
    runtime: str = "unknown"
    container_id: str = "unknown"
    accelerator_type: str = "none"
    accelerator_count: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


# Module-level cache
_cached_context: Optional[DeploymentContext] = None
_cache_expiry: float = 0.0
_CACHE_TTL_SECONDS = 300  # 5 minutes


def detect_deployment_context(*, force: bool = False) -> DeploymentContext:
    """Detect deployment environment from env vars and metadata.

    Cached for 5 minutes. Pass force=True to bypass cache.
    """
    global _cached_context, _cache_expiry

    now = time.monotonic()
    if not force and _cached_context is not None and now < _cache_expiry:
        return _cached_context

    ctx = DeploymentContext()

    # -- Cloud Provider Detection --
    if os.environ.get("AWS_REGION") or os.environ.get("AWS_DEFAULT_REGION"):
        ctx.cloud_provider = "aws"
        ctx.region = os.environ.get("AWS_REGION") or os.environ.get("AWS_DEFAULT_REGION", "unknown")
        ctx.availability_zone = os.environ.get("AWS_AVAILABILITY_ZONE", "unknown")
    elif os.environ.get("GOOGLE_CLOUD_PROJECT") or os.environ.get("GCLOUD_PROJECT"):
        ctx.cloud_provider = "gcp"
        ctx.region = os.environ.get("GOOGLE_CLOUD_REGION", os.environ.get("CLOUD_RUN_REGION", "unknown"))
        ctx.availability_zone = os.environ.get("GOOGLE_CLOUD_ZONE", "unknown")
    elif os.environ.get("AZURE_RESOURCE_GROUP") or os.environ.get("WEBSITE_SITE_NAME"):
        ctx.cloud_provider = "azure"
        ctx.region = os.environ.get("REGION_NAME", os.environ.get("WEBSITE_REGION", "unknown"))
    elif os.environ.get("VULTR_API_KEY") or os.environ.get("VULTR_REGION"):
        ctx.cloud_provider = "vultr"
        ctx.region = os.environ.get("VULTR_REGION", "unknown")

    # -- Runtime Detection --
    if os.environ.get("KUBERNETES_SERVICE_HOST") or os.path.exists("/var/run/secrets/kubernetes.io"):
        ctx.runtime = "kubernetes"
        ctx.container_id = sha256_truncated(os.environ.get("HOSTNAME", "unknown"))
    elif os.environ.get("LAMBDA_TASK_ROOT") or os.environ.get("AWS_LAMBDA_FUNCTION_NAME"):
        ctx.runtime = "lambda"
    elif os.environ.get("ECS_CONTAINER_METADATA_URI") or os.environ.get("ECS_CONTAINER_METADATA_URI_V4"):
        ctx.runtime = "ecs"
        ctx.container_id = sha256_truncated(os.environ.get("HOSTNAME", "unknown"))
    elif os.environ.get("CLOUD_RUN_JOB") or os.environ.get("K_SERVICE"):
        ctx.runtime = "cloud-run"
    elif os.environ.get("AZURE_FUNCTIONS_ENVIRONMENT"):
        ctx.runtime = "azure-functions"
    elif os.path.exists("/.dockerenv") or os.path.exists("/run/.containerenv"):
        ctx.runtime = "container"
        ctx.container_id = sha256_truncated(os.environ.get("HOSTNAME", "unknown"))
    else:
        ctx.runtime = "bare-metal"

    # -- Accelerator Detection (lightweight, no subprocess) --
    if os.environ.get("NVIDIA_VISIBLE_DEVICES") or os.environ.get("CUDA_VISIBLE_DEVICES"):
        devices = os.environ.get("NVIDIA_VISIBLE_DEVICES") or os.environ.get("CUDA_VISIBLE_DEVICES", "")
        if devices and devices != "none":
            ctx.accelerator_type = "nvidia-gpu"
            if devices == "all":
                ctx.accelerator_count = -1  # unknown count, all GPUs
            else:
                ctx.accelerator_count = len(devices.split(","))
    elif os.environ.get("TPU_NAME") or os.environ.get("TPU_WORKER_HOSTNAMES"):
        ctx.accelerator_type = "tpu"
        ctx.accelerator_count = 1
    elif os.environ.get("NEURON_RT_VISIBLE_CORES"):
        ctx.accelerator_type = "aws-neuron"
        cores = os.environ.get("NEURON_RT_VISIBLE_CORES", "")
        ctx.accelerator_count = len(cores.split(",")) if cores else 1

    _cached_context = ctx
    _cache_expiry = now + _CACHE_TTL_SECONDS
    return ctx


def context_to_observations(
    ctx: DeploymentContext,
    clearing_level: int = 1,
) -> Dict[str, Any]:
    """Convert DeploymentContext to observations dict, clearing-level aware.

    CL 0-1: Full context (all fields).
    CL 2: Provider + runtime only (no region/zone/container).
    CL 3: All fields hashed.
    """
    if clearing_level <= 1:
        return ctx.to_dict()
    elif clearing_level == 2:
        return {
            "cloud_provider": ctx.cloud_provider,
            "runtime": ctx.runtime,
            "accelerator_type": ctx.accelerator_type,
            "accelerator_count": ctx.accelerator_count,
        }
    else:
        return {
            "cloud_provider": sha256_truncated(ctx.cloud_provider, 8),
            "runtime": sha256_truncated(ctx.runtime, 8),
            "region": sha256_truncated(ctx.region, 8),
        }


def reset_cache() -> None:
    """Reset the deployment context cache (for testing)."""
    global _cached_context, _cache_expiry
    _cached_context = None
    _cache_expiry = 0.0

File: packages/swt3-ai-ts/src/deployment.ts
/**
 * SWT3 AI Witness SDK -- Deployment Context Detection.
 *
 * Auto-detects cloud provider, region, runtime environment, and
 * accelerator hardware from environment variables.
 *
 * Detection is best-effort and never blocks. Unknown fields remain
 * as "unknown". Container IDs and hostnames are SHA-256 hashed.
 */

import { sha256Truncated } from "./fingerprint.js";
import { existsSync } from "node:fs";

export interface DeploymentContext {
  cloud_provider: string;
  region: string;
  availability_zone: string;
  runtime: string;
  container_id: string;
  accelerator_type: string;
  accelerator_count: number;
}

let _cachedContext: DeploymentContext | null = null;
let _cacheExpiry = 0;
const CACHE_TTL_MS = 300_000; // 5 minutes

/**
 * Detect deployment environment from env vars.
 * Cached for 5 minutes. Pass force=true to bypass cache.
 */
export function detectDeploymentContext(opts?: { force?: boolean }): DeploymentContext {
  const now = Date.now();
  if (!opts?.force && _cachedContext && now < _cacheExpiry) {
    return _cachedContext;
  }

  const env = process.env;
  const ctx: DeploymentContext = {
    cloud_provider: "unknown",
    region: "unknown",
    availability_zone: "unknown",
    runtime: "unknown",
    container_id: "unknown",
    accelerator_type: "none",
    accelerator_count: 0,
  };

  // -- Cloud Provider --
  if (env.AWS_REGION || env.AWS_DEFAULT_REGION) {
    ctx.cloud_provider = "aws";
    ctx.region = env.AWS_REGION || env.AWS_DEFAULT_REGION || "unknown";
    ctx.availability_zone = env.AWS_AVAILABILITY_ZONE || "unknown";
  } else if (env.GOOGLE_CLOUD_PROJECT || env.GCLOUD_PROJECT) {
    ctx.cloud_provider = "gcp";
    ctx.region = env.GOOGLE_CLOUD_REGION || env.CLOUD_RUN_REGION || "unknown";
    ctx.availability_zone = env.GOOGLE_CLOUD_ZONE || "unknown";
  } else if (env.AZURE_RESOURCE_GROUP || env.WEBSITE_SITE_NAME) {
    ctx.cloud_provider = "azure";
    ctx.region = env.REGION_NAME || env.WEBSITE_REGION || "unknown";
  } else if (env.VULTR_API_KEY || env.VULTR_REGION) {
    ctx.cloud_provider = "vultr";
    ctx.region = env.VULTR_REGION || "unknown";
  }

  // -- Runtime --
  if (env.KUBERNETES_SERVICE_HOST) {
    ctx.runtime = "kubernetes";
    ctx.container_id = sha256Truncated(env.HOSTNAME || "unknown");
  } else if (env.LAMBDA_TASK_ROOT || env.AWS_LAMBDA_FUNCTION_NAME) {
    ctx.runtime = "lambda";
  } else if (env.ECS_CONTAINER_METADATA_URI || env.ECS_CONTAINER_METADATA_URI_V4) {
    ctx.runtime = "ecs";
    ctx.container_id = sha256Truncated(env.HOSTNAME || "unknown");
  } else if (env.CLOUD_RUN_JOB || env.K_SERVICE) {
    ctx.runtime = "cloud-run";
  } else if (env.AZURE_FUNCTIONS_ENVIRONMENT) {
    ctx.runtime = "azure-functions";
  } else {
    try {
      if (existsSync("/.dockerenv") || existsSync("/run/.containerenv")) {
        ctx.runtime = "container";
        ctx.container_id = sha256Truncated(env.HOSTNAME || "unknown");
      } else {
        ctx.runtime = "bare-metal";
      }
    } catch {
      ctx.runtime = "bare-metal";
    }
  }

  // -- Accelerator --
  const nvDevices = env.NVIDIA_VISIBLE_DEVICES || env.CUDA_VISIBLE_DEVICES;
  if (nvDevices && nvDevices !== "none") {
    ctx.accelerator_type = "nvidia-gpu";
    ctx.accelerator_count = nvDevices === "all" ? -1 : nvDevices.split(",").length;
  } else if (env.TPU_NAME || env.TPU_WORKER_HOSTNAMES) {
    ctx.accelerator_type = "tpu";
    ctx.accelerator_count = 1;
  } else if (env.NEURON_RT_VISIBLE_CORES) {
    ctx.accelerator_type = "aws-neuron";
    ctx.accelerator_count = env.NEURON_RT_VISIBLE_CORES.split(",").length;
  }

  _cachedContext = ctx;
  _cacheExpiry = now + CACHE_TTL_MS;
  return ctx;
}

/**
 * Convert DeploymentContext to observations dict, clearing-level aware.
 *
 * CL 0-1: Full context.
 * CL 2: Provider + runtime only.
 * CL 3: All fields hashed.
 */
export function contextToObservations(
  ctx: DeploymentContext,
  clearingLevel: number = 1,
): Record<string, unknown> {
  if (clearingLevel <= 1) {
    return { ...ctx };
  } else if (clearingLevel === 2) {
    return {
      cloud_provider: ctx.cloud_provider,
      runtime: ctx.runtime,
      accelerator_type: ctx.accelerator_type,
      accelerator_count: ctx.accelerator_count,
    };
  } else {
    return {
      cloud_provider: sha256Truncated(ctx.cloud_provider, 8),
      runtime: sha256Truncated(ctx.runtime, 8),
      region: sha256Truncated(ctx.region, 8),
    };
  }
}

/** Reset the cache (for testing). */
export function resetCache(): void {
  _cachedContext = null;
  _cacheExpiry = 0;
}
